File: src/main/java/pl/mjedynak/idea/plugins/pit/actions/DirectoryOrFilePitAction.java
package pl.mjedynak.idea.plugins.pit.actions;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.DataKeys;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import pl.mjedynak.idea.plugins.pit.configuration.PitRunConfiguration;

public abstract class DirectoryOrFilePitAction extends PitAction {

    abstract String getTitleForItem(final String item);

    abstract boolean isEnabled(
            @NotNull final Project project, @NotNull final Module module, @NotNull final VirtualFile vfile);

    @Override
    public void update(@NotNull final AnActionEvent e) {
        final Project project = e.getData(DataKeys.PROJECT);
        final Module module = e.getData(DataKeys.MODULE);
        final VirtualFile vfile = e.getData(DataKeys.VIRTUAL_FILE);

        final boolean enabled =
                (project != null) && (module != null) && (vfile != null) && isEnabled(project, module, vfile);

        e.getPresentation().setEnabledAndVisible(enabled);
        if (enabled) {
            e.getPresentation().setText(getTitleForItem(vfile.getPresentableName()));
        }
    }

    abstract PitRunConfiguration makeConfigurationForClassList(
            @NotNull final String classList, @NotNull final Project project, @NotNull final String title);

    @Override
    protected PitRunConfiguration getConfigurationForActionEvent(final AnActionEvent e) {
        final Project project = e.getData(DataKeys.PROJECT);
        final Module module = e.getData(DataKeys.MODULE);
        final VirtualFile vfile = e.getData(DataKeys.VIRTUAL_FILE);

        if ((project == null) || (module == null) || (vfile == null) || !isEnabled(project, module, vfile)) {
            return null;
        }

        final List<String> classNames;
        if (vfile.isDirectory()) {
            final PsiDirectory psiDirectory = PsiManager.getInstance(project).findDirectory(vfile);
            classNames = PitActionUtils.getClassNamesInDirectory(psiDirectory);
        } else {
            final PsiFile psiFile = PsiManager.getInstance(project).findFile(vfile);
            classNames = PitActionUtils.getClassNamesForFile(psiFile);
        }

        if (classNames.isEmpty()) {
            return null;
        }

        final String joinedString = String.join(",", classNames);

        return makeConfigurationForClassList(joinedString, project, vfile.getPresentableName());
    }
}

File: src/main/kotlin/pl/mjedynak/idea/plugins/pit/actions/DirectoryOrFilePitAction.kt
package pl.mjedynak.idea.plugins.pit.actions

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.PlatformCoreDataKeys
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiManager
import pl.mjedynak.idea.plugins.pit.configuration.PitRunConfiguration

abstract class DirectoryOrFilePitAction : PitAction() {
    protected abstract fun getTitleForItem(item: String): String

    protected abstract fun isEnabled(
        project: Project,
        module: Module,
        vfile: VirtualFile,
    ): Boolean

    override fun update(e: AnActionEvent) {
        val project = e.getData(CommonDataKeys.PROJECT)
        val module = e.getData(PlatformCoreDataKeys.MODULE)
        val vfile = e.getData(CommonDataKeys.VIRTUAL_FILE)
        val enabled = project != null && module != null && vfile != null && isEnabled(project, module, vfile)
        e.presentation.isEnabledAndVisible = enabled
        if (project != null && module != null && vfile != null && enabled) {
            e.presentation.text = getTitleForItem(vfile.presentableName)
        }
    }

    protected abstract fun makeConfigurationForClassList(
        classList: String,
        project: Project,
        title: String,
    ): PitRunConfiguration

    override fun getConfigurationForActionEvent(e: AnActionEvent): PitRunConfiguration? {
        val project = e.getData(CommonDataKeys.PROJECT) ?: return null
        val module = e.getData(PlatformCoreDataKeys.MODULE) ?: return null
        val vfile = e.getData(CommonDataKeys.VIRTUAL_FILE) ?: return null
        if (!isEnabled(project, module, vfile)) return null

        val classNames =
            if (vfile.isDirectory) {
                val psiDirectory = PsiManager.getInstance(project).findDirectory(vfile)
                PitActionUtils.getClassNamesInDirectory(psiDirectory)
            } else {
                val psiFile = PsiManager.getInstance(project).findFile(vfile)
                PitActionUtils.getClassNamesForFile(psiFile)
            }

        if (classNames.isEmpty()) return null
        val joinedString = classNames.joinToString(",")
        return makeConfigurationForClassList(joinedString, project, vfile.presentableName)
    }
}
